### PR TITLE
add stock command

### PR DESCRIPTION
**Pull request by ExtraHash**

### CODE REVIEW

File: lib/CommandImplementations.ts [hunk 1]
```diff
@@ -542,6 +542,62 @@ function formatVaccineData(data: any, population: number, embed: MessageEmbed) {
     );
 }
 
+export async function handleStock(msg: Message, args: string[]) {
+    if (args.length === 0) {
+        msg.channel.send("You need to include a ticker. Example: $stock IBM");
```
Comment (zpalmtree) [inline lib/CommandImplementations.ts:547]:
```suggestion
        msg.channel.send(`You need to include a ticker. Example: \`${config.ticker}stock IBM\``);
```

File: lib/CommandImplementations.ts [hunk 2]
```diff
@@ -542,6 +542,62 @@ function formatVaccineData(data: any, population: number, embed: MessageEmbed) {
...
+
+    const [ ticker ] = args;
+
+    const res = await fetch(`https://www.alphavantage.co/query?function=GLOBAL_QUOTE&symbol=${ticker.toUpperCase()}&apikey=${config.stockApiKey}`);
+    const stockData = await res.json();
+
+    const embed = new MessageEmbed()
+    .setColor(Number(stockData['Global Quote']['09. change']) < 0 ? '#C8102E' : '#00853D')
+    .setTitle(ticker.toUpperCase())
+    // .setThumbnail('https://i.imgur.com/FnbQwqQ.png')
```
Comment (zpalmtree) [inline lib/CommandImplementations.ts]:
```suggestion
```

File: lib/CommandImplementations.ts [hunk 3]
```diff
@@ -542,6 +542,62 @@ function formatVaccineData(data: any, population: number, embed: MessageEmbed) {
...
+export async function handleStock(msg: Message, args: string[]) {
+    if (args.length === 0) {
+        msg.channel.send("You need to include a ticker. Example: $stock IBM");
+        return;
+    }
+
+    const [ ticker ] = args;
+
+    const res = await fetch(`https://www.alphavantage.co/query?function=GLOBAL_QUOTE&symbol=${ticker.toUpperCase()}&apikey=${config.stockApiKey}`);
+    const stockData = await res.json();
```
Comment (zpalmtree) [inline lib/CommandImplementations.ts]: We need to check if a response is actually retrieved and warn the user that their ticker is invalid if not.

File: lib/CommandImplementations.ts [hunk 4]
```diff
@@ -542,6 +542,62 @@ function formatVaccineData(data: any, population: number, embed: MessageEmbed) {
...
+            value: `$${numberWithCommas(stockData['Global Quote']['03. high'])}`,
+            inline: true,
+        },
+    );
+
+
+    msg.channel.send(embed);
+}
+
+function numberWithCommas(s: string) {
```
Comment (zpalmtree) [inline lib/CommandImplementations.ts:609]: This should be moved to `Utilities.ts`

File: lib/CommandImplementations.ts [hunk 5]
```diff
@@ -542,6 +542,62 @@ function formatVaccineData(data: any, population: number, embed: MessageEmbed) {
...
+    const stockData = await res.json();
+
+    const embed = new MessageEmbed()
+    .setColor(Number(stockData['Global Quote']['09. change']) < 0 ? '#C8102E' : '#00853D')
+    .setTitle(ticker.toUpperCase())
+    // .setThumbnail('https://i.imgur.com/FnbQwqQ.png')
+    .addFields(
+        {
+            name: 'Price',
+            value: `$${numberWithCommas(stockData['Global Quote']['05. price'])}`,
```
Comment (zpalmtree) [inline lib/CommandImplementations.ts]: We should probably call `roundToNPlaces(val, 2)` for all the values in this embed to cap them at a sensible length.
You might want to alias the formatting, to keep it short, e.g.

```js
const f = (val) => numberWithCommas(formatToNPlaces(val, 2));
```

File: lib/CommandDeclarations.ts [hunk 6]
```diff
@@ -761,6 +762,15 @@ export const Commands: Command[] = [
             },
         ],
     },
+    {
+        aliases: ['stock', 'stonk'],
+        primaryCommand: {
+            argsFormat: Args.Split,
+            implementation: handleStock,
+            description: 'Check a stock price.',
+            needDb: false,
```
Comment (zpalmtree) [inline lib/CommandDeclarations.ts:771]: Can we add an example with a ticker being given. The default generated help does not include a ticker value.

File: lib/CommandDeclarations.ts [hunk 7]
```diff
@@ -761,6 +762,15 @@ export const Commands: Command[] = [
             },
         ],
     },
+    {
+        aliases: ['stock', 'stonk'],
+        primaryCommand: {
+            argsFormat: Args.Split,
+            implementation: handleStock,
+            description: 'Check a stock price.',
```
Comment (zpalmtree) [inline lib/CommandDeclarations.ts:770]:
```suggestion
            description: 'Check a stock price',
```

**Review by zpalmtree — CHANGES_REQUESTED** [review]

**Comment by ExtraHash** [comment]
should be resolved

File: lib/CommandImplementations.ts [hunk 8]
```diff
@@ -542,6 +543,71 @@ function formatVaccineData(data: any, population: number, embed: MessageEmbed) {
...
+    if (args.length === 0) {
+        msg.channel.send(`You need to include a ticker. Example: \`${config.prefix}stock IBM\``);
+        return;
+    }
+
+    const [ ticker ] = args;
+
+    const res = await fetch(`https://www.alphavantage.co/query?function=GLOBAL_QUOTE&symbol=${ticker.toUpperCase()}&apikey=${config.stockApiKey}`);
+
+    if (res.status == 200) {
```
Comment (zpalmtree) [inline lib/CommandImplementations.ts:556]:
```suggestion
    if (res.status === 200) {
```

File: lib/CommandImplementations.ts [hunk 9]
```diff
@@ -542,6 +543,71 @@ function formatVaccineData(data: any, population: number, embed: MessageEmbed) {
...
+        const stockData = await res.json();
+
+        if (Object.entries(stockData['Global Quote']).length == 0) {
```
Comment (zpalmtree) [inline lib/CommandImplementations.ts:559]:
```suggestion
        if (Object.entries(stockData['Global Quote']).length === 0) {
```

**Review by zpalmtree — CHANGES_REQUESTED** [review]
Certain ticker inputs can cause unexpected results, or the function to throw.

For example `$stonk A;FOO` or `$stonk ;`

This maybe wil be fixed by using `stringify` on the query parameters to correctly encode them.

**Comment by ExtraHash** [comment]
fixed

**Pull request merged**